src/lib.rs

```diff
@@ -30,22 +31,41 @@ pub fn load_happs_yaml(path: impl AsRef<Path>) -> Result<Vec<Happ>> {
 }
 
 pub async fn install_happs(happ_list: &[Happ], config: &Config) -> Result<()> {
-    let admin_websocket = AdminWebsocket::connect(config.admin_port)
+    let mut admin_websocket = AdminWebsocket::connect(config.admin_port)
         .await
         .context("failed to connect to holochain")?;
+
+    if let Err(e) = admin_websocket.attach_app_interface(config.happ_port).await {
+        warn!(port = ?config.happ_port, "failed to start interface, maybe it's already up?");
+    }
+
+    let agent_key = admin_websocket
+        .generate_agent_pubkey()
+        .await
+        .context("failed to generate agent key")?;
+
+    let installed_happs = Arc::new(
+        admin_websocket
+            .list_installed_happs()
+            .await
+            .context("failed to get installed hApps")?,
+    );
+
     let futures: Vec<_> = happ_list
         .iter()
         .map(|happ| {
             let mut admin_websocket = admin_websocket.clone();
+            let agent_key = agent_key.clone();
+            let installed_happs = Arc::clone(&installed_happs);
             async move {
-                let mut agent_websocket = admin_websocket.clone();
-                let install_happ = agent_websocket
-                    .generate_agent_pubkey()
-                    .and_then(|agent_key| {
-                        admin_websocket.install_happ(happ, agent_key, config.happ_port)
-                    });
                 let install_ui = install_ui(happ, config);
-                futures::try_join!(install_happ, install_ui)
+                if installed_happs.contains(&happ.id_with_version()) {
```

Comment:
`admin_websocket.install_happ()` is not atomic, it goes in two stages - installApp() and activateApp(), therefore  technically this check should verify if an app is both installed AND active, and act accordingly. I am fine though for now to leave it as a TODO item :-)